lib/logstash/inputs/tcp.rb

```diff
@@ -86,48 +83,121 @@ def register
         @ssl_context.cert_store = @cert_store
         @ssl_context.verify_mode = OpenSSL::SSL::VERIFY_PEER|OpenSSL::SSL::VERIFY_FAIL_IF_NO_PEER_CERT
       end
-    end # @ssl_enable
+    end
+
+    # note that since we are opening a socket in register, we must also make sure we close it
+    # in the close method even if we also close it in the stop method since we could have
+    # a situation where register is called but not run & stop.
 
     if server?
       @logger.info("Starting tcp input listener", :address => "#{@host}:#{@port}")
       begin
-        @server_socket = TCPServer.new(@host, @port)
+        set_server_socket(TCPServer.new(@host, @port))
       rescue Errno::EADDRINUSE
         @logger.error("Could not start TCP server: Address in use", :host => @host, :port => @port)
         raise
       end
-      if @ssl_enable
-        @server_socket = OpenSSL::SSL::SSLServer.new(@server_socket, @ssl_context)
-      end # @ssl_enable
+
+      set_server_socket(OpenSSL::SSL::SSLServer.new(server_socket, @ssl_context)) if @ssl_enable
+    end
+  end
+
+  def run(output_queue)
+    if server?
+      run_server(output_queue)
+    else
+      run_client(output_queue)
     end
-  end # def register
+  end
+
+  def stop
+    # force close all sockets which will escape any blocking read with a IO exception
+    # and any thread using them will exit
+    server_socket.close rescue nil
+    client_socket.close rescue nil
+    connection_sockets.each{|socket| socket.close rescue nil}
+  end
+
+  def close
+    # see related comment in register: we must make sure to close the server socket here
+    # because it is created in the register method and we could be in the context of having
+    # register called but never run & stop, only close.
+    server_socket.close rescue nil
+  end
 
   private
-  def handle_socket(socket, client_address, output_queue, codec)
-    while !@interrupted
-      buf = nil
-      # NOTE(petef): the timeout only hits after the line is read or socket dies
-      # TODO(sissel): Why do we have a Timeout here? What's the point?
-      if @data_timeout == -1
-        buf = read(socket)
-      else
-        Timeout::timeout(@data_timeout) do
-          buf = read(socket)
+
+  def run_server(output_queue)
+    while !stop?
+      begin
+        socket = add_connection_socket(server_socket.accept)
+        # start a new thread for each connection.
+        server_connection_thread(output_queue, socket)
+      rescue OpenSSL::SSL::SSLError => e
+        @logger.error("SSL Error", :exception => e, :backtrace => e.backtrace)
+      rescue
+        # if this exception occured while the plugin is stopping
+        # just ignore and exit
+        raise unless stop?
+      end
+    end
+  ensure
+    server_socket.close rescue nil
+  end
+
+  def run_client(output_queue)
+    while !stop?
+      set_client_socket(TCPSocket.new(@host, @port))
+
+      if @ssl_enable
```

Comment:
It could I guess, and a lot of code could be better refactored here too. tried to keep it to minimal.